lib/spack/llnl/util/cpu.py

```diff
@@ -0,0 +1,316 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
```

Comment:
Call this `llnl.util.archspec`, or even move it to `external/archspec.py`, as we'd like to use that name for it eventually.

Reply:
Would it be worth to keep `cpu.py` under `archspec` in case we extend later to other aspects (like network, or detection of accellerators, etc.)?

Reply:
Yeah I think that's reasonable.  Though we should think about how people are going to import/write this stuff.  Is it:

```python
import archspec.cpu as cpu
arch = cpu.detect_march()
```

The modules should somehow reflect the usage.

Reply:
@tgamblin @becker33 I'll change the layout of the code on the filesystem last. Would you agree with a structure like:
```console
archspec/cpu
├── __init__.py
├── alias.py
├── detection.py
└── microarchitecture.py
```
where `__init__.py` imports names for the most commonly used APIs from submodules? In that way a  user can do things like:
```python
import archspec.cpu as cpu
from archspec.cpu import MicroArchitecture
```
and we will hide by default the more advanced APIs (like the `alias_predicate` decorator, see 28b2d96). In other words `__init__.py` will be a sort of "facade" to wrap the complexity of submodules and expose a simple interface.

Also, for the time being I'd leave the code under `llnl.util` mainly because of the unit tests that are currently part of `spack/test`. I think that as long as this is not factored in its own package (i.e. another repository under the Spack organization) it would be clearer to have it there.

As soon as this code is moved into another repository with the corresponding unit tests we can then:
- Vendor it as external
- Remove the unit tests from Spack

What do you think?

Reply:
I think this is a good plan -- the layout looks good to me.